bluebinder.c

```diff
@@ -368,38 +369,49 @@ bluebinder_callbacks_transact(
     if (!g_strcmp0(iface, BINDER_BLUETOOTH_SERVICE_IFACE_CALLBACKS)) {
         if (code == 1) { /* initializationComplete */
             int result = 0;
+            GBinderLocalReply* reply = NULL;
 
             gbinder_remote_request_read_int32(req, &result);
 
-            if (result != 0) {
-                fprintf(stderr, "Bluetooth binder service failed\n");
-                g_main_loop_quit(proxy->loop);
+            do{
+
                 *status = GBINDER_STATUS_FAILED;
-                return NULL;
-            }
+                proxy->is_bt_failed_on_init = TRUE;
 
-            fprintf(stderr, "Binder interface initialized, opening virtual device\n");
+                if (result != 0) {
+                    fprintf(stderr, "Bluetooth binder service failed\n");
+                    /* we need to tell BT service that we properly handled Status::INITIALIZATION_ERROR */
+                    *status = GBINDER_STATUS_OK;
+                    reply = gbinder_local_reply_append_int32(gbinder_local_object_new_reply(obj), 0);
+                    break;
```

Comment:
I'm not a fan of these breaks that don't break anything. It's essentially a goto. @krnlyng are you ok with that?

Reply:
i have actually never seen a do {} while(0) with a break i believe, looks very much like a goto, i agree, i think we can avoid that

Reply:
Thats a common practice, guys, it is replacement for goto

Static analisys tools and future maintainers will say thank to you ;)

Reply:
It's not a replacement for goto, it's (almost) an equivalent of goto. I agree that it's slightly better than a classic goto (because these "unnamed" gotos must be nested) and in that sense it's a step forward, but I personally prefer straight linear logic without any gotos at all, named or unnamed.

If I were a maintainer, I would definitely insist on not using this sort of programming, which IMO turns a program into a riddle. But I don't consider myself a maintainer of this piece of code, so I'm just casually pointing it out.

Reply:
i would prefer not having the implicit goto

Reply:
Should I replace do {} while(0) with multiple returns? (like it was before)

Reply:
BTW do {} while(0) is very common when you want to pass SonarQube tests, also it is very common when you use open ssl. Fo ex: https://wiki.openssl.org/index.php/EVP_Signing_and_Verifying

This construction allows to avoid goto and keep code flow strict and predictive. 

As for me, I also don't like goto at all, so lets replace all of that with just many returns. That it is also possible.

Reply:
I personally would find something like this easier to read and understand:
```c
    if (bluebinder_init_complete(proxy)) {
        *status = GBINDER_STATUS_OK;
        return gbinder_local_reply_append_int32(gbinder_local_object_new_reply(obj), 0);
    } else {
        proxy->init_failed = TRUE;
        *status = GBINDER_STATUS_FAILED;
        return NULL;
    }
```
Two returns (success and failure) would be enough, and I think it's nice when they are close enough to each other to see them both on the same screen.

And of course there are many ways to do the same thing, many different patterns, most of which make certain amount of sense from certain point of view. Somebody has to make an arbitrary choice which one to apply.

Reply:
ok, I have refactored the code a little bit